pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+				basic_asset_price *
+					(T::MyUnsignedFixedPoint::from(T::SellFee::get()) +
+						T::MyUnsignedFixedPoint::one())
+			};
+			let exchange_asset_price =
+				oracle::Pallet::<T>::get_price(OracleKey::ExchangeRate(asset))?.into();
+
+			let exchange_amount = Self::multiply_by_rational(
+				buyout_amount.into(),
+				basic_asset_price_with_fee.into_inner(),
+				exchange_asset_price.into_inner(),
+			)
+			.map(|n| n.try_into().ok())
+			.flatten()
+			.ok_or(ArithmeticError::Overflow.into());
+
+			exchange_amount
+		}
+
+		fn calc_buyout_amount(
+			asset: CurrencyIdOf<T>,
+			exchange_amount: BalanceOf<T>,
+		) -> Result<BalanceOf<T>, DispatchError> {
+			let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+
+			ensure!(asset != basic_asset, Error::<T>::WrongAssetToBuyout,);
+
+			//TODO fix this
+			let basic_asset_price_with_fee = {
+				let basic_asset_price =
+					oracle::Pallet::<T>::get_price(OracleKey::ExchangeRate(basic_asset))?.into();
+				(basic_asset_price *
+					(T::MyUnsignedFixedPoint::from(T::SellFee::get()) +
+						T::MyUnsignedFixedPoint::one()))
+				.into_inner()
+			};
+			let exchange_asset_price =
+				oracle::Pallet::<T>::get_price(OracleKey::ExchangeRate(asset))?.into();
+
+			let buyout_amount = Self::multiply_by_rational(
+				exchange_amount.into(),
+				exchange_asset_price,
+				basic_asset_price_with_fee,
+			)
+			.map(|b| b.try_into().ok())
+			.flatten()
+			.ok_or(ArithmeticError::Overflow.into());
+
+			buyout_amount
+		}
+
+		fn split_to_buyout_and_exchange(
+			asset: CurrencyIdOf<T>,
+			amount: Amount<BalanceOf<T>>,
+		) -> Result<(BalanceOf<T>, BalanceOf<T>), DispatchError> {
+			match amount {
+				Amount::Buyout(buyout_amount) => {
+					let exchange_amount = Self::calc_amount_to_exchange(asset, buyout_amount)?;
+					Ok((buyout_amount, exchange_amount))
+				},
+				Amount::Exchange(exchange_amount) => {
+					let buyout_amount = Self::calc_buyout_amount(asset, exchange_amount)?;
+					Ok((buyout_amount, exchange_amount))
+				},
+			}
+		}
+
+		fn do_buyout(
+			who: AccountIdOf<T>,
+			asset: CurrencyIdOf<T>,
+			amount: Amount<BalanceOf<T>>,
+		) -> DispatchResult {
+			Self::ensure_not_native_buyout(&asset)?;
+			let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+			let (buyout_amount, exchange_amount) =
+				Self::split_to_buyout_and_exchange(asset, amount)?;
+			Self::ensure_buyout_limit_not_exceeded(&who, buyout_amount)?;
+			let treasury_account_id = T::TreasuryAccount::get();
+
+			Self::exchange(
+				(&who, &treasury_account_id),
+				(&asset, &basic_asset),
+				(exchange_amount, buyout_amount),
+			)
+			.map_err(|(error, maybe_acc)| match maybe_acc {
+				Some(acc) =>
+					if acc == treasury_account_id {
+						Error::<T>::InsufficientTreasuryBalance.into()
+					} else if acc == who {
+						Error::<T>::InsufficientAccountBalance.into()
+					} else {
+						error
+					},
+				_ => error,
+			})?;
```

Comment:
Maybe it's best if we don't extract the logic for 'exchanging' the tokens into an extra function. It's done like this by equilibrium as they implement this as a primitive on their balances pallet. But to be honest, I'm not a huge fan of that function overall, and I find it to be more confusing than helping with readability.

So I think it's best if instead you put the logic that is now hidden in the `exchange()` function to here. Then we can remove the `ExchangeFailure` error type and we also don't need to deposit the `Exchange` event IMO as calling the `transfer()` function on the `Currencies` trait will emit similar events anyways.